cpp/src/arrow/type_test.cc

```diff
@@ -379,6 +382,72 @@ TEST(TestFieldPath, Basics) {
                                   FieldPath({s.num_fields() * 2}).Get(s));
 }
 
+TEST(TestFieldPath, GetForTable) {
+  using testing::HasSubstr;
+
+  const int length = 100;
+  auto f0 = field("alpha", int32());
+  auto f1 = field("beta", int32());
+  auto f2 = field("alpha", int32());
+  auto f3 = field("beta", int32());
+  auto schema = arrow::schema({f0, f1, f2, f3});
+
+  arrow::random::RandomArrayGenerator gen_{42};
+  auto a0 = gen_.ArrayOf(int32(), length);
+  auto a1 = gen_.ArrayOf(int32(), length);
+  auto a2 = gen_.ArrayOf(int32(), length);
+  auto a3 = gen_.ArrayOf(int32(), length);
+  auto arrayVector = ArrayVector({a0, a1, a2, a3});
+
+  auto tablePtr = Table::Make(schema, {a0, a1, a2, a3});
```

Comment:
It would be worth throwing in a `ASSERT_OK(tablePtr->ValidateFull())` after this (same for `RecordBatch::Make` i guess).